{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Single-cell technologies have revolutionized biology by allowing us to study the transcriptome, epigenome, or proteome of individual cells, rather than averaging over bulk populations. However, single-cell datasets are inherently high-dimensional, sparse, and noisy. Each cell is represented by thousands of features (genes), but many of these features are zero in any given cell due to dropout events. Machine learning (ML) offers powerful tools to handle these challenges: dimensionality reduction, clustering, and predictive modeling can reveal hidden patterns, trajectories, and gene programs in single-cell data.\n",
    "\n",
    "In this tutorial, we will explore the workflow of analyzing single-cell transcriptomic data using ML."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data is from https://www.10xgenomics.com/datasets/pbmc-from-a-healthy-donor-no-cell-sorting-3-k-1-standard-2-0-0, containing the PBMC 3k of ATAC and expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 3009 × 36601\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'interval'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/homebrew/Caskroom/miniconda/base/envs/BioML-tut-llm/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/opt/homebrew/Caskroom/miniconda/base/envs/BioML-tut-llm/lib/python3.10/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "h5_file = \"data_singlecell/pbmc_unsorted_3k_filtered_feature_bc_matrix.h5\"\n",
    "\n",
    "adata = sc.read_10x_h5(h5_file)\n",
    "\n",
    "adata.var_names_make_unique()\n",
    "\n",
    "print(adata)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3009, 36601)\n"
     ]
    }
   ],
   "source": [
    "print(adata.X.shape) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "\n",
    "# Read fragments\n",
    "fragments = pd.read_csv(\n",
    "    \"data_singlecell/pbmc_unsorted_3k_atac_fragments.tsv.gz.tsv\",\n",
    "    sep='\\t',\n",
    "    header=None,\n",
    "    names=['chrom','start','end','barcode','count','strand'],\n",
    "    compression='gzip',\n",
    "    comment='#'\n",
    ")\n",
    "\n",
    "# Filter for RNA cells\n",
    "cell_barcodes = set(adata.obs_names)\n",
    "fragments = fragments[fragments['barcode'].isin(cell_barcodes)]\n",
    "\n",
    "# Rename columns for PyRanges\n",
    "fragments = fragments.rename(columns={\n",
    "    'chrom': 'Chromosome',\n",
    "    'start': 'Start',\n",
    "    'end': 'End',\n",
    "    'strand': 'Strand'  # optional\n",
    "})\n",
    "\n",
    "# Create PyRanges object\n",
    "fr = pr.PyRanges(fragments[['Chromosome','Start','End','barcode','count','Strand']])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "\n",
    "# Read peaks BED, skipping comment lines\n",
    "peaks = pd.read_csv(\n",
    "    \"data_singlecell/pbmc_unsorted_3k_atac_peaks.bed\",\n",
    "    sep='\\t',\n",
    "    header=None,\n",
    "    names=['chrom','start','end'],\n",
    "    comment='#'   # <-- skip any lines starting with #\n",
    ")\n",
    "\n",
    "# Rename columns for PyRanges\n",
    "peaks = peaks.rename(columns={\n",
    "    'chrom': 'Chromosome',\n",
    "    'start': 'Start',\n",
    "    'end': 'End'\n",
    "})\n",
    "\n",
    "# Now create PyRanges object\n",
    "pk = pr.PyRanges(peaks)\n",
    "overlaps = fr.join(pk)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>barcode</th>\n",
       "      <th>count</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start_b</th>\n",
       "      <th>End_b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL000009.2</td>\n",
       "      <td>114064</td>\n",
       "      <td>114295</td>\n",
       "      <td>GACCTCAAGGAACGGT-1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113930</td>\n",
       "      <td>114840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GL000009.2</td>\n",
       "      <td>114247</td>\n",
       "      <td>114474</td>\n",
       "      <td>GCTAGCCAGGCGAATA-1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113930</td>\n",
       "      <td>114840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GL000009.2</td>\n",
       "      <td>114267</td>\n",
       "      <td>114440</td>\n",
       "      <td>TATTCGTTCAATCTCT-1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113930</td>\n",
       "      <td>114840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>28130</td>\n",
       "      <td>28352</td>\n",
       "      <td>TACGCACCAATAGCAA-1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27950</td>\n",
       "      <td>28827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>28146</td>\n",
       "      <td>28381</td>\n",
       "      <td>GTCCATTGTGGATGTC-1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27950</td>\n",
       "      <td>28827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19178517</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56847220</td>\n",
       "      <td>56847315</td>\n",
       "      <td>GAAGGAACATTCAGCA-1</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>56846844</td>\n",
       "      <td>56847757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19178518</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56847423</td>\n",
       "      <td>56847562</td>\n",
       "      <td>GTTGGCGGTTCGCGCT-1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>56846844</td>\n",
       "      <td>56847757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19178519</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56847435</td>\n",
       "      <td>56847577</td>\n",
       "      <td>AGTAATCGTGGATGTC-1</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>56846844</td>\n",
       "      <td>56847757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19178520</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56847512</td>\n",
       "      <td>56847679</td>\n",
       "      <td>GACCTCAAGGAACGGT-1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>56846844</td>\n",
       "      <td>56847757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19178521</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56847577</td>\n",
       "      <td>56847629</td>\n",
       "      <td>GAAGGAACATTCAGCA-1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>56846844</td>\n",
       "      <td>56847757</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19178522 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------------+-------+\n",
       "| Chromosome   | Start     | End       | barcode            | +4    |\n",
       "| (category)   | (int64)   | (int64)   | (object)           | ...   |\n",
       "|--------------+-----------+-----------+--------------------+-------|\n",
       "| GL000009.2   | 114064    | 114295    | GACCTCAAGGAACGGT-1 | ...   |\n",
       "| GL000009.2   | 114247    | 114474    | GCTAGCCAGGCGAATA-1 | ...   |\n",
       "| GL000009.2   | 114267    | 114440    | TATTCGTTCAATCTCT-1 | ...   |\n",
       "| GL000194.1   | 28130     | 28352     | TACGCACCAATAGCAA-1 | ...   |\n",
       "| ...          | ...       | ...       | ...                | ...   |\n",
       "| GL000194.1   | 28206     | 28356     | GTTAACCAGTTTGCGG-1 | ...   |\n",
       "| GL000194.1   | 28285     | 28360     | GTCCAAGTCTCCATGC-1 | ...   |\n",
       "| GL000009.2   | 114064    | 114295    | GACCTCAAGGAACGGT-1 | ...   |\n",
       "| chrY         | 56847423  | 56847562  | GTTGGCGGTTCGCGCT-1 | ...   |\n",
       "| chrY         | 56847435  | 56847577  | AGTAATCGTGGATGTC-1 | ...   |\n",
       "| chrY         | 56847512  | 56847679  | GACCTCAAGGAACGGT-1 | ...   |\n",
       "| chrY         | 56847577  | 56847629  | GAAGGAACATTCAGCA-1 | ...   |\n",
       "+--------------+-----------+-----------+--------------------+-------+\n",
       "Unstranded PyRanges object has 19,178,522 rows and 8 columns from 39 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome.\n",
       "Considered unstranded due to these Strand values: 'nan'\n",
       "4 hidden columns: count, Strand, Start_b, End_b"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "import numpy as np\n",
    "from scipy.sparse import coo_matrix\n",
    "import scanpy as sc\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Chromosome   Start     End             barcode  count Strand  Start_b  \\\n",
      "0  GL000009.2  114064  114295  GACCTCAAGGAACGGT-1      1    NaN   113930   \n",
      "1  GL000009.2  114247  114474  GCTAGCCAGGCGAATA-1      1    NaN   113930   \n",
      "2  GL000009.2  114267  114440  TATTCGTTCAATCTCT-1      1    NaN   113930   \n",
      "3  GL000194.1   28130   28352  TACGCACCAATAGCAA-1      1    NaN    27950   \n",
      "4  GL000194.1   28146   28381  GTCCATTGTGGATGTC-1      1    NaN    27950   \n",
      "\n",
      "    End_b  \n",
      "0  114840  \n",
      "1  114840  \n",
      "2  114840  \n",
      "3   28827  \n",
      "4   28827  \n",
      "Index(['Chromosome', 'Start', 'End', 'barcode', 'count', 'Strand', 'Start_b',\n",
      "       'End_b'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(overlaps.df.head())\n",
    "print(overlaps.df.columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign peak indices\n",
    "peaks_df = pk.df.reset_index(drop=True)\n",
    "peaks_df['peak_idx'] = np.arange(len(peaks_df))\n",
    "\n",
    "# Merge overlaps with peak indices\n",
    "overlaps_df = overlaps.df.merge(\n",
    "    peaks_df[['Chromosome','Start','End','peak_idx']],\n",
    "    left_on=['Chromosome','Start_b','End_b'],  # notice Chromosome alone\n",
    "    right_on=['Chromosome','Start','End'],\n",
    "    how='left'\n",
    ")\n",
    "\n",
    "# Map barcodes to RNA cell indices\n",
    "cell_to_idx = {cell:i for i, cell in enumerate(adata.obs_names)}\n",
    "overlaps_df['cell_idx'] = overlaps_df['barcode'].map(cell_to_idx)\n",
    "\n",
    "# Build sparse matrix\n",
    "from scipy.sparse import coo_matrix\n",
    "\n",
    "rows = overlaps_df['cell_idx'].to_numpy()\n",
    "cols = overlaps_df['peak_idx'].to_numpy()\n",
    "data = overlaps_df['count'].to_numpy()\n",
    "\n",
    "X_atac = coo_matrix((data, (rows, cols)), shape=(adata.n_obs, len(peaks_df)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3009, 81156)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_atac.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<COOrdinate sparse matrix of dtype 'int64'\n",
       "\twith 19178522 stored elements and shape (3009, 81156)>"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/homebrew/Caskroom/miniconda/base/envs/BioML-tut-llm/lib/python3.10/site-packages/anndata/_core/storage.py:48: FutureWarning: AnnData previously had undefined behavior around matrices of type <class 'scipy.sparse._coo.coo_matrix'>.In 0.12, passing in this type will throw an error. Please convert to a supported type.Continue using for this minor version at your own risk.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/opt/homebrew/Caskroom/miniconda/base/envs/BioML-tut-llm/lib/python3.10/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    }
   ],
   "source": [
    "import scipy.sparse as sp\n",
    "import scanpy as sc\n",
    "atac_adata = sc.AnnData(X=X_atac, obs=adata.obs.copy(), var=peaks)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((3009, 81156), (3009, 36601))"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac_adata.shape, adata.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 3009 × 81156\n",
      "    var: 'Chromosome', 'Start', 'End'\n"
     ]
    }
   ],
   "source": [
    "print(atac_adata)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BioML-tut-llm",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
